lib/TPP/Transforms/FoldAddIntoDest.cpp

```diff
@@ -0,0 +1,110 @@
+//===- FoldAddIntoDest.cpp ---------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Passes.h"
+#include "TPP/Transforms/Utils/ValueUtils.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/Interfaces/DestinationStyleOpInterface.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+using namespace mlir;
+
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_FOLDADDINTODEST
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+
+namespace {
+
+/// Replace a linalg.add where its linalg-contraction operand - with a
+/// zero-filled destination - is dominated by the `other` linalg operand,
+/// by passing `other` as the destination of the contraction.
+struct FoldAddIntoDestRewrite : public OpRewritePattern<linalg::AddOp> {
```

Comment:
I think the add can't be folded when the contraction has other users. Consider this example:
```
%0 = linalg.sub %A, %B
%1 = linalg.matmul %C, %D, %zero
%2 = linalg.add %0, %1
%3 = linalg.mul %1, %2
```

Reply:
Good catch. This is fixed now, with a test case.